src/loadsave.jl

```diff
@@ -1,37 +1,119 @@
+    path = Base.find_in_path(string(pkg)) # hacky way to determine if a Package is installed
+    path == nothing && return false
+    return isfile(path)
+end
+
+function save_import(pkg::Symbol)
+    !is_installed(pkg)      && throw(NotInstalledError(pkg, ""))
+    !isdefined(Main, pkg)   && eval(Main, Expr(:import, pkg))
+    return Main.(pkg)
+end
+
+
+for (applicable_, add_, dict_) in (
+        (:applicable_loaders, :add_loader, :sym2loader),
+        (:applicable_savers,  :add_saver,  :sym2saver))
     @eval begin
-        $appl{sym}(::Formatted{DataFormat{sym}}) = get($dct, sym, [:FileIO]) # if no loader is declared, fallback to FileIO
-        function $fchk(pkg::Symbol)
-            !isdefined(Main, pkg) && eval(Main, Expr(:import, pkg))
-            return Main.(pkg)
-        end
-        function $fadd{sym}(::Type{DataFormat{sym}}, pkg::Symbol)
-            list = get($dct, sym, Symbol[])
-            $dct[sym] = push!(list, pkg)
+        $applicable_{sym}(::@compat(Union{Type{DataFormat{sym}}, Formatted{DataFormat{sym}}})) = get($dict_, sym, [:FileIO]) # if no loader is declared, fallback to FileIO
+        function $add_{sym}(::Type{DataFormat{sym}}, pkg::Symbol)
+            list = get($dict_, sym, Symbol[])
+            $dict_[sym] = push!(list, pkg)
         end
     end
 end
 
-applicable_mime{sym}(::MIME{sym}) = get(mimedict, sym, [:nothing])
-function check_mime(pkg::Symbol)
-    pkg == :nothing && error("No MIME package available")
-    !isdefined(Main, pkg) && eval(Main, Expr(:import, pkg))
-    return pkg
-end
+applicable_mime{sym}(::MIME{sym}) = get(mimedict, sym, [(Any, :FileIO)])
 
 @doc """
 `add_mime(mime, T, :Package)`  triggers `using Package` before attempting to write object of type `T` in format `mime`.
 """ ->
 function add_mime{sym,T}(::MIME{sym}, ::Type{T}, pkg::Symbol)
-    list = get(mimedict, sym, Any[])
+    list = get(mimedict, sym, @compat(Tuple{DataType, Symbol})[])
     mimedict[sym] = push!(list, (T,pkg))
 end
 
 @doc "`add_loader(fmt, :Package)` triggers `using Package` before loading format `fmt`" -> add_loader
 @doc "`add_saver(fmt, :Package)` triggers `using Package` before saving format `fmt`" -> add_saver
+
+
+@doc """
+- `load(filename)` loads the contents of a formatted file, trying to infer
+the format from `filename` and/or magic bytes in the file.
+- `load(strm)` loads from an `IOStream` or similar object. In this case,
+the magic bytes are essential.
+- `load(File(format"PNG",filename))` specifies the format directly, and bypasses inference.
+- `load(f; options...)` passes keyword arguments on to the loader.
+""" ->
+function load(s::@compat(Union{AbstractString,IO}), args...; options...)
+    q           = query(s)
+    libraries   = applicable_loaders(q)
+    failures    = Any[]
+    for library in libraries
+        try
+            Library = save_import(library)
+            return Library.load(q, args...; options...)
```

Comment:
Yes, I'm in favor of it just being `using FileIO`, and the user not having to worry about which package gets used. (`using Images` is necessary only if the user also wants to call functions in `Images`.)

I guess part of my issue here is that, if we use @SimonDanisch's strategy, there is absolutely no reason to make the format a type-parameter: we're basically "dispatching" by dictionary-lookup and try/catch, in which case `DataFormat` should be defined
```jl
immutable DataFormat
    format::Symbol
end
```
so we reduce our compile burden.

My original intention was that dispatch would be direct, once you knew what format you wanted (and users could control that format with `format"FMT"`) and the relevant package was loaded. In particular, you could load or save a file without needing to, e.g., check what package folders are present in the user's path each time you want to load another file of the exact same format. But there are problems, chief among them that there can be only one exported `load(f::File{format"FMT"})` function, and the last one to be loaded always wins (and no module-scoping can change that). The "escape hatch" I envisioned was the unexported `load_` mechanism discussed in #16 (it really would have been better if that PR had gotten more serious discussion earlier on...), and even that's only partial because it doesn't solve the "oops, I just replaced `QuartzImageIO`'s PNG-loader with one from `ImageMagick`". Just about the only solution I see to that problem is to have ImageMagick be considerate and not stomp on previous versions if they are already defined. For example:
```
function __init__()
    if !isdefined(Main, :QuartzImageIO)   # or could use !applicable(load, File(format"PNG", "dummy")) as a general sentinel?
        eval(Expr(:import, :IMBindings))
    end
end
```
The internal `IMBindings` module would define the globally-exported versions,
```
FileIO.load(...) = ImageMagick.load_(...)
```
You can see a more complete example this kind of design in ProfileView (see its interaction with ProfileViewGtk through their respective `__init__` functions). If it seems a little complicated, blame package precompilation and the fact that you can't count on the user's system having the same set of packages loaded at usage time as s/he did at compile time.

So, I think we're faced with a choice:
- Allow direct dispatch of the type I originally designed: FileIO gets out of the way once the appropriate package is already loaded, except possibly to query the magic bytes when the user doesn't specify the format directly. The downside is that life gets considerably more complicated for developers in situations like the QuartzImageIO/ImageMagick format conflict, both in avoiding stomping on each others' methods and also in knowing what to do if one package has an error and would rather defer to another one.
- Switch to Simon's "dispatch"-by-dictionary-lookup-and-try/catch. While arguably less julian, it does have the major advantage of not needing any kind of cooperation among "competing" loaders/savers---it relies on FileIO to know "everything," so individual loaders/savers can just throw an error when they get confused.

Right now we have an awkward hybrid of these two strategies. Let's try to decide on a course of action soon, and get on with it. I can get on board with your strategy, @SimonDanisch, but if we do that we should start from the ground up (with the definition of `DataFormat`).